library/init/meta/async_tactic.lean

```diff
@@ -0,0 +1,49 @@
+prelude
+import init.meta.tactic
+import init.meta.interactive
+
+namespace tactic
+
+private meta def run_or_fail {α} (s : tactic_state) (tac : tactic α) : α :=
+match tac s with
+| (tactic_result.success a s) := a
+| (tactic_result.exception .α fmt _ s') :=
+  undefined_core $ to_string $ fmt () ++ format.line ++ to_fmt s'
+end
+
+meta def run_async {α : Type} (tac : tactic α) : tactic (task α) := do
+s ← read, return $ task.delay $ λ _,
+  match tac s with
+  | (tactic_result.success a s) := a
+  | (tactic_result.exception .α fmt _ s') :=
+    undefined_core $ to_string $ fmt () ++ format.line ++ to_fmt s'
+  end
+
+meta def prove_goal_async (tac : tactic unit) : tactic unit := do
+ctx ← local_context, revert_lst ctx,
+tgt ← target, tgt ← instantiate_mvars tgt,
+env ← get_env, tgt ← return $ env^.unfold_untrusted_macros tgt,
+when tgt^.has_meta_var (fail $ "goal contains metavariables"),
+params ← return tgt^.collect_univ_params,
+lemma_name ← mk_fresh_name,
```

Comment:
`mk_fresh_name` is not good for creating lemma names. It only guarantees freshness in the current session. For example, suppose we use `async` in files `A.lean` and `B.lean`. Then, when we import `A` and `B`, there is a risk of name collision.

One possible ugly solution is to pass the name `C` of surrounding declaration, and create a name such as `C._aux_<idx>`. We can make it less ugly by storing the name of the surrounding declaration in the `tactic_state`.